libyear-gradle-plugin/src/main/kotlin/com/libyear/traversal/DependencyTraversal.kt

```diff
@@ -37,12 +45,50 @@ class DependencyTraversal private constructor(
     }
   }
 
+  @VisibleForTesting
+  fun shouldIncludeModule(
+    module: String,
+    depth: Int
+  ): Boolean {
+    if (maxTransitiveDepth != null && depth > maxTransitiveDepth) {
+      return false
+    }
+
+    // Inclusions supersede exclusions
+    val matchedInclusion = includedPatterns.firstOrNull { pattern -> pattern.second.matches(module) }
+    val matchedExclusion = excludedPatterns.firstOrNull { pattern -> pattern.second.matches(module) }
+
+    if (matchedInclusion != null) {
+      logger.info("Including $module because it matches ${matchedInclusion.first}")
+    } else if (matchedExclusion != null) {
+      logger.info("Excluding $module because it matches ${matchedExclusion.first}")
+      return false
+    }
+
+    return true
```

Comment:
I think we do have a misunderstanding from https://github.com/f4lco/libyear-gradle-plugin/pull/25#discussion_r1818039345 🙈 The point of the example was to show that the substring logic is too fuzzy to be useful. The slf4j example transferred badly to inclusion / exclusion lists.

I think the inclusion list is supposed to "narrow down" on a subset of dependencies, and supposed to answer questions like "how many libyears does the Spring framework bring in?", or perhaps "how many Libyears do dependencies from `com.mycompany:*` bring in?".

Drawing my inspiration from https://ant.apache.org/manual/dirtasks.html:

> Only include files and directories that match any include patterns and do not match any exclude patterns
> 
> When both inclusion and exclusion are used, only files/directories that match at least one of the include patterns and don't match any of the exclude patterns are used.
> If no include pattern is given, all files are assumed to match the include pattern

I think that's how most (build) tools work. Please find an implementation sketch below:

```kotlin
val matchedInclusion = includedPatterns.firstOrNull { pattern -> pattern.second.matches(module) }
val matchedExclusion = excludedPatterns.firstOrNull { pattern -> pattern.second.matches(module) }

val matchesInclusions = includedPatterns.isEmpty() || matchedInclusion != null
val matchesExclusions = matchedExclusion != null
val shouldIncludeModule = matchesInclusions && !matchesExclusions

if (!shouldIncludeModule) {
  if (matchesExclusions) {
    logger.info("Excluding $module because it matches ${matchedExclusion!!.first}")
  } else {
    logger.info("Excluding $module because it does not match inclusion list")
  }
}

return shouldIncludeModule
```

Do you think that is workable?

Reply:
Works for me! I've updated the branch an merged in develop to get the latest other changes (ended up rebasing + force push, sorry if you were tracking the branch too).

Reply:
Nudge nudge @f4lco

Reply:
> (ended up rebasing + force push, sorry if you were tracking the branch too).

No worries!